pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -480,6 +512,265 @@ func containerSucceeded(c *v1.Container, podStatus *kubecontainer.PodStatus) boo
+		}
+	}
+	resizeMemLim, restartMemLim := determineContainerResize(v1.ResourceMemory, desiredMemoryLimit, currentMemoryLimit)
+	resizeCPULim, restartCPULim := determineContainerResize(v1.ResourceCPU, desiredCPULimit, currentCPULimit)
+	resizeCPUReq, restartCPUReq := determineContainerResize(v1.ResourceCPU, desiredCPURequest, currentCPURequest)
+	if restartCPULim || restartCPUReq || restartMemLim {
+		// resize policy requires this container to restart
+		changes.ContainersToKill[kubeContainerStatus.ID] = containerToKillInfo{
+			name:      kubeContainerStatus.Name,
+			container: &pod.Spec.Containers[containerIdx],
+			message:   fmt.Sprintf("Container %s resize requires restart", container.Name),
+		}
+		changes.ContainersToStart = append(changes.ContainersToStart, containerIdx)
+		changes.UpdatePodResources = true
+		return false
+	} else {
+		if resizeMemLim {
+			markContainerForUpdate(v1.ResourceMemory, desiredMemoryLimit, currentMemoryLimit)
+		}
+		if resizeCPULim {
+			markContainerForUpdate(v1.ResourceCPU, desiredCPULimit, currentCPULimit)
+		} else if resizeCPUReq {
+			markContainerForUpdate(v1.ResourceCPU, desiredCPURequest, currentCPURequest)
+		}
+	}
+	return true
+}
+
+func (m *kubeGenericRuntimeManager) doPodResizeAction(pod *v1.Pod, podStatus *kubecontainer.PodStatus, podContainerChanges podActions, result kubecontainer.PodSyncResult) {
+	pcm := m.containerManager.NewPodContainerManager()
+	//TODO(vinaykul,InPlacePodVerticalScaling): Figure out best way to get enforceMemoryQoS value (parameter #4 below) in platform-agnostic way
+	podResources := cm.ResourceConfigForPod(pod, m.cpuCFSQuota, uint64((m.cpuCFSQuotaPeriod.Duration)/time.Microsecond), false)
+	if podResources == nil {
+		klog.ErrorS(nil, "Unable to get resource configuration", "pod", pod.Name)
+		result.Fail(fmt.Errorf("Unable to get resource configuration processing resize for pod %s", pod.Name))
+		return
+	}
+	setPodCgroupConfig := func(rName v1.ResourceName, setLimitValue bool) error {
+		var err error
+		switch rName {
+		case v1.ResourceCPU:
+			podCpuResources := &cm.ResourceConfig{CpuPeriod: podResources.CpuPeriod}
+			if setLimitValue == true {
+				podCpuResources.CpuQuota = podResources.CpuQuota
+			} else {
+				podCpuResources.CpuShares = podResources.CpuShares
+			}
+			err = pcm.SetPodCgroupConfig(pod, rName, podCpuResources)
+		case v1.ResourceMemory:
+			err = pcm.SetPodCgroupConfig(pod, rName, podResources)
+		}
+		if err != nil {
+			klog.ErrorS(err, "Failed to set cgroup config", "resource", rName, "pod", pod.Name)
+		}
+		return err
+	}
+	// Memory and CPU are updated separately because memory resizes may be ordered differently than CPU resizes.
+	// If resize results in net pod resource increase, set pod cgroup config before resizing containers.
+	// If resize results in net pod resource decrease, set pod cgroup config after resizing containers.
+	// If an error occurs at any point, abort. Let future syncpod iterations retry the unfinished stuff.
```

Comment:
I am considering this scenario:
One pod (P1) has two containers, C1 and C3
```
P1: 
- C1: CPU limit -> 1 core
- C2: CPU limit -> 2 core

Total: 3 core 
```
we want to resize it to 

```
P1: 
- C1: CPU limit -> 4 core
- C2: CPU limit -> 1 core

Total:  5 core
```

So we will first resize PodCgroup from 3 to 5 and then resize container C1 from 1 to 4, but this must be failed, because we only have a total 5 core limit and can not tolerate the 4+2=6 core transition state right?

Reply:
Whether the lowering resizing container should be prioritized?

Reply:
@qmloong that is correct. We have explicitly captured it in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#container-resource-limit-update-ordering) and implemented it as such. I also plan to go over this point at KubeCon next month.

Reply:
Oh, thanks. I found the sorting logic in `computePodResizeAction`.

Reply:
> @qmloong that is correct. We have explicitly captured it in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources#container-resource-limit-update-ordering) and implemented it as such. I also plan to go over this point at KubeCon next month.

@vinaykul could you tell me which session it is?

Reply:
@bg-chun The session is https://sched.co/182HU on Wed Oct 26.
While the focus of this talk is about how we can leverage eBPF to proactively initiate in-place resize, I will have a few slides to cover the design and expectations from CRI & runtimes.

Reply:
resolving conversation.